File: Remote/Source/Bank/BankObject/BankObject Subclasses/ComponentDevice.m
//
// ComponentDevice.m
// Remote
//
// Created by Jason Cardwell on 6/1/11.
// Copyright (c) 2011 Moondeer Studios. All rights reserved.
//
#import "ComponentDevice.h"
#import "IRCode.h"
#import "Command.h"

@implementation ComponentDevice {
    BOOL _ignoreNextPowerCommand;
    BOPowerState _power;
    BODevicePort _port;
}

@dynamic name, port, codes, power, inputPowersOn, alwaysOn, offCommand, onCommand;

/*
+ (id)MR_importFromObject:(id)objectData inContext:(NSManagedObjectContext *)context;
{
    NSAttributeDescription * primaryAttribute = [[self MR_entityDescription]
                                                 MR_primaryAttributeToRelateBy];

    id value = [objectData MR_valueForAttribute:primaryAttribute];

    NSManagedObject * managedObject = [self MR_findFirstByAttribute:[primaryAttribute name]
                                                          withValue:value
                                                          inContext:context];
    if (managedObject == nil)
    {
        managedObject = [self MR_createInContext:context];
    }

    [managedObject MR_importValuesForKeysWithObject:objectData];

    return managedObject;
}
*/

+ (BOOL)isEditable { return NO; }

+ (BOOL)isPreviewable { return NO;}

+ (NSString *)directoryLabel { return @"Component Device"; }

+ (NSOrderedSet *)directoryItems { return nil; }

- (UIImage *)thumbnail { return nil; }

- (UIImage *)preview { return nil; }

- (NSString *)category { return nil; }

- (UIViewController *)editingViewController { return nil; }

- (NSOrderedSet *)subBankables { return nil; }

+ (instancetype)fetchDeviceWithName:(NSString *)name
{
    return [self MR_findFirstByAttribute:@"name" withValue:name];
}

+ (instancetype)fetchDeviceWithName:(NSString *)name context:(NSManagedObjectContext *)context
{
    return [self MR_findFirstByAttribute:@"name" withValue:name inContext:context];
}

- (BOOL)ignorePowerCommand:(RECommandCompletionHandler)handler
{
    if (_ignoreNextPowerCommand)
    {
        _ignoreNextPowerCommand = NO;
        if (handler) handler(YES, nil);
        return YES;
    }

    else return NO;
}

- (void)powerOn:(RECommandCompletionHandler)completion
{
    __weak ComponentDevice * weakself = self;
    if (![self ignorePowerCommand:completion])
        [self.onCommand execute:^(BOOL success, NSError * error) {
            weakself.power = (!error && success ? BOPowerStateOn : BOPowerStateOff);
            if (completion) completion(success, error);
        }];
}

- (void)powerOff:(RECommandCompletionHandler)completion
{
    __weak ComponentDevice * weakself = self;
    if (![self ignorePowerCommand:completion])
        [self.offCommand execute:^(BOOL success, NSError * error) {
            weakself.power = (!error && success ? BOPowerStateOff : BOPowerStateOn);
            if (completion) completion(success, error);
        }];
}

- (void)setPower:(BOPowerState)power
{
    [self willChangeValueForKey:@"power"];
    _power = power;
    [self didChangeValueForKey:@"power"];
}

- (BOPowerState)power
{
    [self willAccessValueForKey:@"power"];
    BOPowerState power = _power;
    [self didAccessValueForKey:@"power"];
    return power;
}

- (void)setPort:(BODevicePort)port
{
    [self willChangeValueForKey:@"port"];
    _port = port;
    [self didChangeValueForKey:@"port"];
}

- (BODevicePort)port
{
    [self willAccessValueForKey:@"port"];
    BODevicePort port = _port;
    [self didAccessValueForKey:@"port"];
    return port;
}

- (IRCode *)objectForKeyedSubscript:(NSString *)name
{
    return [self.codes objectPassingTest:^BOOL(IRCode * obj){return [name isEqualToString:obj.name];}];
}

- (MSDictionary *)deepDescriptionDictionary
{
    ComponentDevice * device = [self faultedObject];
    assert(device);

    MSMutableDictionary * dd = [[super deepDescriptionDictionary] mutableCopy];
    dd[@"name"] = device.name;
    dd[@"port"] = $(@"%i",device.port);
    return dd;
}

@end

File: Projects/MSRemote/Remote/Remote/Source/Bank/Bankables/ComponentDevice.swift
//
//  ComponentDevice.swift
//  Remote
//
//  Created by Jason Cardwell on 9/30/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import CoreData
import MoonKit

@objc(ComponentDevice)
class ComponentDevice: BankableModelObject {

  @NSManaged var alwaysOn: Bool
  @NSManaged var inputPowersOn: Bool
  @NSManaged var primitiveInputs: NSSet
  var inputs: [IRCode] {
    get {
      willAccessValueForKey("inputs")
      let inputs = primitiveInputs.allObjects as? [IRCode]
      didAccessValueForKey("inputs")
      return inputs ?? []
    }
    set {
      willChangeValueForKey("inputs")
      primitiveInputs = NSSet(array: newValue)
      didChangeValueForKey("inputs")
    }
  }
  @NSManaged var port: Int16
  @NSManaged var power: Bool
  @NSManaged var primitiveCodeSet: IRCodeSet?
  var codeSet: IRCodeSet? {
    get {
      willAccessValueForKey("codeSet")
      let codeSet = primitiveCodeSet
      didAccessValueForKey("codeSet")
      return codeSet
    }
    set {
      willChangeValueForKey("codeSet")
      primitiveCodeSet = newValue
      didChangeValueForKey("codeSet")
      if let codeSet = primitiveCodeSet {
        if let manufacturer = primitiveManufacturer {
          if manufacturer != codeSet.manufacturer {
            self.manufacturer = nil
          }
        }
      }
    }
  }
//  @NSManaged var manufacturer: Manufacturer?
  @NSManaged var primitiveManufacturer: Manufacturer?
  var manufacturer: Manufacturer? {
    get {
      willAccessValueForKey("manufacturer")
      let manufacturer = primitiveManufacturer
      didAccessValueForKey("manufacturer")
      return manufacturer
    }
    set {
      willChangeValueForKey("manufacturer")
      primitiveManufacturer = newValue
      didChangeValueForKey("manufacturer")
      if let manufacturer = primitiveManufacturer {
        if let codeSet = primitiveCodeSet {
          if codeSet.manufacturer != manufacturer {
            self.codeSet = nil
          }
        }
      }
    }
  }
  @NSManaged var networkDevice: NetworkDevice?
  @NSManaged var offCommand: SendIRCommand?
  @NSManaged var onCommand: SendIRCommand?
  @NSManaged var powerCommands: NSSet

  private var ignoreNextPowerCommand = false

  func ignorePowerCommand(completion: ((Bool, NSError?) -> Void)?) -> Bool {
    if ignoreNextPowerCommand {
      ignoreNextPowerCommand = false
      completion?(true, nil)
      return true
    } else { return false }
  }


  func powerOn(completion: ((Bool, NSError?) -> Void)?) {
    if !ignorePowerCommand(completion) {
      offCommand?.execute{[unowned self] (success: Bool, error: NSError?) in
        if success { self.power = true }
        completion?(success, error)
      }
    }
  }

  func powerOff(completion: ((Bool, NSError?) -> Void)?) {
    if !ignorePowerCommand(completion) {
      offCommand?.execute{[unowned self] (success: Bool, error: NSError?) in
        if success { self.power = false }
        completion?(success, error)
      }
    }
  }

  /**
  objectForKeyedSubscript:

  :param: name String!

  :returns: AnyObject!
  */
//  subscript(name: String) -> AnyObject! { return (codes.allObjects as [IRCode]).filter{$0.name == name}.first }

  /**
  updateWithData:

  :param: data [NSObject AnyObject]!
  */
  override func updateWithData(data: [NSObject : AnyObject]!) {
    super.updateWithData(data)

    port = (data["port"] as? NSNumber)?.shortValue ?? port
    onCommand = SendIRCommand.importObjectFromData(data["on-command"] as? NSDictionary,
                                           context: managedObjectContext!) ?? onCommand
    offCommand = SendIRCommand.importObjectFromData(data["off-command"] as? NSDictionary,
                                            context: managedObjectContext!) ?? offCommand
    manufacturer = Manufacturer.importObjectFromData(data["manufacturer"] as? NSDictionary,
                                             context: managedObjectContext!) ?? manufacturer
    networkDevice = NetworkDevice.importObjectFromData(data["network-device"] as? NSDictionary,
                                               context: managedObjectContext!) ?? networkDevice
    codeSet = IRCodeSet.importObjectFromData(data["code-set"] as? NSDictionary,
                                     context: managedObjectContext) ?? codeSet
  }

  class var rootCategory: Bank.RootCategory {
    let devices = findAllSortedBy("name", ascending: true) as? [ComponentDevice]
    return Bank.RootCategory(label: "Component Devices",
                             icon: UIImage(named: "969-television")!,
                             items: devices ?? [],
                             editableItems: true)
  }

  /**
  isEditable

  :returns: Bool
  */
  override class func isEditable() -> Bool { return true  }

  /**
  isPreviewable

  :returns: Bool
  */
  override class func isPreviewable() -> Bool { return false }

  /**
  detailController

  :returns: UIViewController
  */
  override func detailController() -> UIViewController {
    return ComponentDeviceDetailController(item: self)!
  }

}

extension ComponentDevice: MSJSONExport {

  /**
  JSONDictionary

  :returns: MSDictionary!
  */
  override func JSONDictionary() -> MSDictionary! {

    let dictionary = super.JSONDictionary()

    setIfNotDefault("port",          inDictionary: dictionary)
    setIfNotDefault("alwaysOn",      inDictionary: dictionary)
    setIfNotDefault("inputPowersOn", inDictionary: dictionary)

    safeSetValueForKeyPath("onCommand.JSONDictionary",    forKey: "on-command",          inDictionary: dictionary)
    safeSetValueForKeyPath("offCommand.JSONDictionary",   forKey: "off-command",         inDictionary: dictionary)
    safeSetValueForKeyPath("manufacturer.commentedUUID",  forKey: "manufacturer.uuid",   inDictionary: dictionary)
    safeSetValueForKeyPath("networkDevice.commentedUUID", forKey: "network-device.uuid", inDictionary: dictionary)
    safeSetValueForKeyPath("codeSet.commentedUUID",       forKey: "code-set",            inDictionary: dictionary)

    dictionary.compact()
    dictionary.compress()

    return dictionary

  }

}
